dosfstools/man/dosfslabel.8
.\" dosfslabel.8 - manpage for dosfsck
.\"
.\" Copyright (C) 2006-2013 Daniel Baumann <daniel@debian.org>
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program. If not, see <http://www.gnu.org/licenses/>.
.\"
.\" On Debian systems, the complete text of the GNU General Public License
.\" can be found in /usr/share/common-licenses/GPL-3 file.
.\"
.\"
.TH DOSFSLABEL 8 "2010\-01\-31" "3.0.9" "set or get MS\-DOS filesystem label"

.SH NAME
\fBdosfslabel\fR \- set or get MS\-DOS filesystem label

.SH SYNOPSIS
\fBdosfslabel\fR \fIDEVICE\fR [\fILABEL\fR]

.SH DESCRIPTION
\fBdosfslabel\fR set or gets a MS\-DOS filesystem label from a given device.
.PP
If the label is omitted, then the label name of the specified device is written on the standard output. A label can't be longer than 11 characters.

.SH OPTIONS
.IP "\fB\-h\fR, \fB\-\-help\fR" 4
Displays a help message.
.IP "\fB\-V\fR, \fB\-\-version\fR" 4
Shows version.

.SH SEE ALSO
\fBdosfsck\fR(8)
.br
\fBmkdosfs\fR(8)

.SH HOMEPAGE
More information about \fBdosfsck\fR and \fBdosfstools\fR can be found at <\fIhttp://www.daniel\-baumann.ch/software/dosfstools/\fR>.

.SH AUTHORS
\fBdosfstools\fR were written by Werner Almesberger <\fIwerner.almesberger@lrc.di.epfl.ch\fR>, Roman Hodek <\fIRoman.Hodek@informatik.uni-erlangen.de\fR>, and others. The current maintainer is Daniel Baumann <\fImail@daniel-baumann.ch\fR>.


dosfstools/ChangeLog
commit ac4d30646425e19e8682cf7c9055014cbdb54c6f
Author: Petr Gajdos <pgajdos@suse.cz>
Date:   Fri Mar 1 08:34:12 2013 +0100

    Create rootdir entry volume label with mkdosfs, create it when
    it doesn't exist with dosfslabel.
    
    See https://bugzilla.novell.com/show_bug.cgi?id=657011#c4
    for more information.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit 7963bda123aa3bc6e4525a69b8fb82c4e92f7a52
Author: Petr Gajdos <pgajdos@suse.cz>
Date:   Fri Mar 1 08:33:18 2013 +0100

    Forbid lowercase letters in label.
    
    See https://bugzilla.novell.com/show_bug.cgi?id=657011#c4 and
    http://support.microsoft.com/kb/71715/en-us for more information.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit 23fd2911267f73ea8dd6b5e6cc7ffb60223063e5
Author: Petr Gajdos <pgajdos@suse.cz>
Date:   Fri Mar 1 08:32:02 2013 +0100

    Read label also from rootdir entry.
    
    See https://bugzilla.novell.com/show_bug.cgi?id=657011#c4
    for more information.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit dbf11935d610b3b8d50480fc6c6257f8d8508635
Author: Petr Gajdos <pgajdos@suse.cz>
Date:   Fri Mar 1 08:30:21 2013 +0100

    alloc_rootdir_entry() is intended to be called with pattern == "FSCK%04dREC",
    the old code (probably c&p from auto_rename()) doesn't reflect this.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit 6de7543008fb5b37017560448c45270dafb62ab8
Author: Petr Gajdos <pgajdos@suse.cz>
Date:   Fri Mar 1 08:29:00 2013 +0100

    Instead of eleven blanks, fill in "NO NAME    " as specification tells.
    
    See https://bugzilla.novell.com/show_bug.cgi?id=657011#c4 and
    http://www.win.tue.nl/~aeb/linux/fs/fat/fat-1.html for more information.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit feb90914727acd92e4518881a07c3d80d6d6ccfa
Author: Daniel Baumann <daniel.baumann@progress-linux.org>
Date:   Fri Mar 1 08:58:15 2013 +0100

    Write uppercase letters in label.
    
    See https://bugzilla.novell.com/show_bug.cgi?id=657011#c4 and
    http://support.microsoft.com/kb/71715/en-us for more information.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit a75fb1c838a47b8f54ff567ca2eb71e98ccb691c
Author: Daniel Baumann <mail@daniel-baumann.ch>
Date:   Thu Feb 21 15:06:52 2013 +0100

    Releasing version 3.0.15.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit c8f84fdcddf0982b86ef53250b08a005c80925ef
Author: Alexander Korolkov <alexander.korolkov@gmail.com>
Date:   Mon Feb 4 00:22:34 2013 +0400

    Using wcstombs() to convert LFN unicode characters to printable text.
    
    This closes Debian bug #596336.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit 154654555730874c1656bb6c63292a85a2260bdf
Author: Alexander Korolkov <alexander.korolkov@gmail.com>
Date:   Sun Sep 5 18:59:47 2010 +0400

    Recode short filenames from DOS codepage (default 437).
    
    Recode short filenames from DOS codepage (default 437) to the current
    character encoding. This makes messages of dosfsck more readable.
    Partially closes Debian bug #596336.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit ad76cae648d3b7895e97e16f8212ae8089b8846b
Author: Jaroslav Skarvada <jskarvad@redhat.com>
Date:   Thu Feb 21 14:40:52 2013 +0100

    Fixing root directory allocation.
    
    See https://bugzilla.redhat.com/show_bug.cgi?id=674095 for more information.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit b8201b34eb83c2bf87bd1c2b1a7fa57a1dc28191
Author: Jaroslav Skarvada <jskarvad@redhat.com>
Date:   Thu Feb 21 14:40:25 2013 +0100

    Fixing device detection.
    
    See https://bugzilla.redhat.com/show_bug.cgi?id=710480 for more information.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit 7a756385ed6bce393396c699e068b5f239053db6
Author: Daniel Baumann <mail@daniel-baumann.ch>
Date:   Wed Jan 23 13:22:01 2013 +0100

    Releasing version 3.0.14.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit 5bdd7ef95c7f18e9e6571bf0718a564bd15b30d5
Author: Daniel Baumann <mail@daniel-baumann.ch>
Date:   Wed Jan 23 13:16:20 2013 +0100

    Documenting dosfsck -b in its manpage.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit a307be2b9cbfd0ed8c10dab79044cf6c7426ea18
Author: Oleksij Rempel <bug-track@fisher-privat.net>
Date:   Wed Jan 23 12:36:56 2013 +0100

    Adding option for bootsector read-only check.
    
    Most boot sectors may contains marker for filesystem state. We can this
    bit on every mount and warn user if some thing wrong, without checking
    complete filesystem.
    
    Signed-off-by: Oleksij Rempel <bug-track@fisher-privat.net>
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit ce2f8dc349a8a6b2c26d1b3892cc9b5f0408bd34
Author: Oleksij Rempel <bug-track@fisher-privat.net>
Date:   Wed Jan 23 12:35:13 2013 +0100

    Checking boot sector for dirty bit.
    
    Some OSos use reseved byte of boot sector to set state of the file
    system. If first bit set, then filesystem is proably damaged - write
    operation was not finished/cache not snycted/...
    
    Signed-off-by: Oleksij Rempel <bug-track@fisher-privat.net>
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit f33ee8ca5640191e7666402d6676073712b9c988
Author: Daniel Baumann <mail@daniel-baumann.ch>
Date:   Wed Jan 23 12:25:59 2013 +0100

    Completing and updating all copyright headers for 2013.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit bfe6d25c6297b2fbca174ead56dffee340668a06
Author: Daniel Baumann <mail@daniel-baumann.ch>
Date:   Wed Jan 23 12:17:20 2013 +0100

    Updating my email address.
    
    Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>

commit 13cdb4d262b4042ec2d94d706e30f45a63f1c029
Author: Daniel Baumann <mail@daniel-baumann.ch>

dosfstools/src/dosfslabel.c
/* dosfslabel.c - User interface

   Copyright (C) 1993 Werner Almesberger <werner.almesberger@lrc.di.epfl.ch>
   Copyright (C) 1998 Roman Hodek <Roman.Hodek@informatik.uni-erlangen.de>
   Copyright (C) 2007 Red Hat, Inc.
   Copyright (C) 2008-2013 Daniel Baumann <mail@daniel-baumann.ch>

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program. If not, see <http://www.gnu.org/licenses/>.

   On Debian systems, the complete text of the GNU General Public License
   can be found in /usr/share/common-licenses/GPL-3 file.
*/

#include "version.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <ctype.h>

#include "common.h"
#include "dosfsck.h"
#include "io.h"
#include "boot.h"
#include "fat.h"
#include "file.h"
#include "check.h"

int interactive = 0, rw = 0, list = 0, test = 0, verbose = 0, write_immed = 0;
int atari_format = 0;
unsigned n_files = 0;
void *mem_queue = NULL;

static void usage(int error)
{
    FILE *f = error ? stderr : stdout;
    int status = error ? 1 : 0;

    fprintf(f, "usage: dosfslabel device [label]\n");
    exit(status);
}

/*
 * ++roman: On m68k, check if this is an Atari; if yes, turn on Atari variant
 * of MS-DOS filesystem by default.
 */
static void check_atari(void)
{
#ifdef __mc68000__
    FILE *f;
    char line[128], *p;

    if (!(f = fopen("/proc/hardware", "r"))) {
	perror("/proc/hardware");
	return;
    }

    while (fgets(line, sizeof(line), f)) {
	if (strncmp(line, "Model:", 6) == 0) {
	    p = line + 6;
	    p += strspn(p, " \t");
	    if (strncmp(p, "Atari ", 6) == 0)
		atari_format = 1;
	    break;
	}
    }
    fclose(f);
#endif
}

int main(int argc, char *argv[])
{
    DOS_FS fs = {0};
    rw = 0;

    int i;

    char *device = NULL;
    char *label = NULL;

    loff_t offset;
    DIR_ENT de;

    check_atari();

    if (argc < 2 || argc > 3)
	usage(1);

    if (!strcmp(argv[1], "-h") || !strcmp(argv[1], "--help"))
	usage(0);
    else if (!strcmp(argv[1], "-V") || !strcmp(argv[1], "--version")) {
	printf("dosfslabel " VERSION ", " VERSION_DATE ", FAT32, LFN\n");
	exit(0);
    }

    device = argv[1];
    if (argc == 3) {
	label = argv[2];
	if (strlen(label) > 11) {
	    fprintf(stderr,
		    "dosfslabel: labels can be no longer than 11 characters\n");
	    exit(1);
	}
        for (i = 0; i < 11; i++)
          /* don't know if here should be more strict !uppercase(label[i])*/
          if (islower(label[i])) {
            fprintf(stderr,
                    "dosfslabel: labels cannot contain lower case characters\n");
            exit(1);
          }
	rw = 1;
    }

    fs_open(device, rw);
    read_boot(&fs);
    if (fs.fat_bits == 32)
	read_fat(&fs);
    if (!rw) {
        offset = find_volume_de(&fs, &de);
        if (offset == 0)
          fprintf(stdout, "%s\n", fs.label);
        else
          fprintf(stdout, "%.8s%.3s\n", de.name, de.ext);
	exit(0);
    }

    write_label(&fs, label);
    fs_close(rw);
    return 0;
}
